### DISCUSSION TITLE
Isolated tests running in their own transactions

### DISCUSSION THREAD

**Discussion by OskarPersson** [tokio-rs/axum#759] [Q&A] [answered]
Looking at the [sqlx-postgres example](https://github.com/tokio-rs/axum/blob/main/examples/sqlx-postgres/src/main.rs) as well as the [testing example](https://github.com/tokio-rs/axum/blob/main/examples/testing/src/main.rs), it feels like it should be possible to have the same app and handlers accept either a sqlx Pool or Transaction object as the extension layer depending on context.

In my case I would like to use a pool as the extension layer and have it be extracted to a connection in the handler (just as in the sqlx-postgres example) when running the app. But when testing, I would like to be able to use the same exact app and handler, but send it a transaction object instead where multiple entries in the database might have been created before making the request such that the handler also works inside the **same** transaction object as the one in the test.

This way I can test against a real database but have each test isolated in their own transaction. I'm having problems achieving this, is this something that could be done?

**Comment by davidpdrsn** [comment, marked as answer]
I don't think thats possible since extractors (things that implement `FromRequest`) most be fully owned. That is, they cannot contain borrowed data. [`sqlx::Transaction`](https://docs.rs/sqlx/latest/sqlx/struct.Transaction.html) borrows the connection so it cannot be accessed via an extractor. I'm not very familiar with sqlx so dunno if there is a workaround.

Also, in general I would strongly recommend against wrapping each test in a transaction. I've attempted to do something like that in the past but kept running into issues, like the one you're facing. Instead I would recommend giving each test its own totally separate database. This is how I do it at work and it works great. We basically have some test helpers that creates a database with a random name, runs the migrations, and automatically removes the database when the test ends (via a `Drop` implementation). It might seem kinda "heavy" to create separate databases for each test but our tests are still plenty fast.

**Reply by OskarPersson** [reply]
Yeah that's kind of what I'm doing now as well. It's just that I wish it could go somewhat faster and its also what I'm used to coming from Django which does this [built-in](https://docs.djangoproject.com/en/4.0/topics/testing/tools/#django.test.TestCase).

Could you provide a snippet of the test helpers you mentioned and how you use them? I would be interested mainly in how you handle the case when a test fails, how is the database deleted then if that code is never accessed?

**Reply by davidpdrsn** [reply]
It's closed source so cannot share it unfortunately.

But it generates a random string with rand and issues some sql commands to create the database and run the migrations. Then you make an [RAII guard](https://rust-unofficial.github.io/patterns/patterns/behavioural/RAII.html) that removes the database in its `Drop` impl. `Drop` is still called for test failures (because they're panics) so that just works.

**Reply by OskarPersson** [reply]
Hmm okay thanks. Is the Drop implemented on the pool object?

**Reply by davidpdrsn** [reply]
No. You make your own struct that contains the name of the database and implement `Drop` for that.

**Reply by OskarPersson** [reply]
I see. Well I'll see if I can figure it out. An example of this part somewhere would be greatly appreciated though. But I understand if that's something that can be hard to extract from the source code :)

**Reply by davidpdrsn** [reply]
Yeah a bit. I would also like to keep the examples focused on specific integrations with axum. This is kinda in a gray area. It would be almost identical to the current sqlx/tokio-postgres examples except have more setup code which isn't really related to axum.

**Reply by OskarPersson** [reply]
Yeah I didn't mean in the repo. Just in this conversation or as a github gist or something :)
